OCPBUGS-7621-overhual: Overhauled the Configuring an external load ba… (#65501)
openshift/openshift-docs: modules/nw-osp-configuring-external-load-balancer.adoc

endif::[]
to use an external load balancer in place of the default load balancer.

[IMPORTANT]
====

> Configuring an external load balancer depends on a specific vendor's load balancer. 

Reviewer: I think this might read better if it said "your specific vendor's load balancer" 

and consult your vendors info...etcetera

+
[TIP]
====
You can list all IP addresses that exist in a network by checking the machine config pool's resources.
====

.Considerations

* For a front-end IP address, you could use the same IP address for the front-end IP address, the Ingress Controller's load balancer and API load balancer. Check the vendor's documentation for this capability.

* For a backend IP address, ensure that an IP address for an {product-title} control plane node does not change during the lifetime of the external load balancer. You can achieve by completing one of the following actions:
** Assign a static IP address to each control plane node.
** Configure each node to receive the same IP address from the DHCP every time the node requests a DHCP lease. Depending on the vendor, the DHCP lease might be in the form of an IP reservation or a static DHCP assignment. 

* For the Ingress Controller backend service, ensure you manually define each node that runs the Ingress Controller in the external load balancer. For example, if the Ingress Controller moves to an undefined node, an connection outage can occur. 

.OpenShift API prerequisites

* You defined a front-end IP address.

> * You ensured that on your load balancer, TCP ports 6443, 443, 22623, and 80 are exposed on the front-end IP address. Check the following items:

Reviewer: For the api/machine config API, we can remove 443 and 80.  We only need those in the ingress controller section.

Author: Thanks. Updated!

** Port 6443 provides access to the OpenShift API service.
** Port 22623 can provide ignition startup configurations to nodes. 
* The front-end IP address and port 6443 are reachable by all users of your system with a location external to your {product-title} cluster.
* The front-end IP address and port 22623 are reachable only by {product-title} nodes.
* The load balancer backend can communicate with {product-title} control plane nodes on port 6443 and 22623.

.Ingress Controller prerequisites

* You defined a front-end IP address.
* You ensured that on your load balancer, TCP ports 443 and 80 are exposed on the front-end IP address.
* The front-end IP address, port 80 and port 443 are be reachable by all users of your system with a location external to your {product-title} cluster. 
* The front-end IP address, port 80 and port 443 are reachable to all nodes that operate in your {product-title} cluster.
* The load balancer backend can communicate with {product-title} nodes that run the Ingress Controller on ports 80, 443, and 1936.

> .Procedure

Reviewer: Instead of procedure, maybe "Configuration Example?"

Author: Hi Clark. The procedure title is essential for a procedure-type module. If I change the module to a concept type, I would need to remove a lot of the headings and remove the action steps "you must, and so on". WDYT?

Reviewer: Rereading through, I think Procedure is OK.  

What's confusing on reading this through is the first step "Configure Health Check Resources for your backend services.

I think if we move Line 89 - 123 into a section _before_ the Procedure.  Called "Health Check URL Specifications"

Reviewer: "Most load balancers allow setting health check URLs to determine if a services is available or not.  OpenShift provides these for the OpenShift API, Machine Configuration API, and Ingress Controller.

Author: Hi @clarkhale . Just wondering about "OpenShift provides these for the OpenShift API, Machine Configuration API, and Ingress Controller." . Does OCP provide load balancers, health checks, or services for the listed APIs?

OpenShift API, Machine Configuration API, and Ingress Controller = services?

Reviewer: > Does OCP provide load balancers, health checks, or services for the listed APIs?
Health checks

> OpenShift API, Machine Configuration API, and Ingress Controller = backend services?

Collectively, yes, in context of the load balancer.

Author: Thanks, Clark. Updated.

. Configure health check resources for your backend services.
+

> .Example of a Kubernetes API server resource

Reviewer: With my above comment,  change this to "OpenShift API Health Check Specification"

+
[source,yaml]
----
Path: HTTPS:6443/readyz
Healthy threshold: 2
Unhealthy threshold: 2
Timeout: 10
Interval: 10
----
+

> .Example of a Machine config server resource

Reviewer: With my above comment, change this to "Machine Config API Health Check Specification"

+
[source,yaml]
----
Path: HTTPS:22623/healthz
Healthy threshold: 2
Unhealthy threshold: 2
Timeout: 10
Interval: 10
----
+

> .Example of an Ingress Controller resource for HTTP and HTTPS traffic

Reviewer: With my above comment, change this to "Ingress Controller Health Check Specification"

endif::[]
to use an external load balancer in place of the default load balancer.

[IMPORTANT]
====
Configuring an external load balancer depends on your vendor's load balancer.

The information and examples in this section are for guideline purposes only. Consult the vendor documentation for more specific information about the vendor's load balancer.
====

Red Hat supports the following services for an external load balancer:

* OpenShift API
* Ingress Controller

> You can choose to configure one or both of these services for an external load balancer, with configuring only the Ingress Controller service as a common configuration route.

Reviewer 2: ```suggestion
You can choose to configure one or both of these services for an external load balancer. Configuring only the Ingress Controller service is a common configuration option.
```
Suggest breaking up sentence and minor edits for readability

Reviewer 2: Suggest using "option" throughout instead of "route" - might just be me, but when I read "route" in these docs I think of [these types of routes](https://docs.openshift.com/container-platform/4.13/networking/routes/route-configuration.html) and that may be confusing for some.

> The following routes exist for configuring an external load balancer:

Reviewer 2: ```suggestion
The following configuration options are supported for external load balancers:
```

> * Use a node selector to map the Ingress Controller to a specific set of nodes, with  infrastructure nodes being a common node type that receives this configuration. You must then assign a static IP address to each node in this set, or configure each node to receive the same IP address from the Dynamic Host Configuration Protocol (DHCP).
> * Target all IP addresses on a subnet. This configuration can reduce maintenance overhead,because you can create and destroy nodes within those networks without reconfiguring the load balancer targets. If you deploy your ingress pods by using a machine set on a smaller network, such as a `/27` or `/28`, you can simplify your load balancer targets.

Reviewer 2: ```suggestion
* Use a node selector to map the Ingress Controller to a specific set of nodes. Then, you must assign a static IP address to each node in this set, or configure each node to receive the same IP address from the Dynamic Host Configuration Protocol (DHCP). Infrastructure nodes commonly receive this type of configuration.

* Target all IP addresses on a subnet. This configuration can reduce maintenance overhead, because you can create and destroy nodes within those networks without reconfiguring the load balancer targets. If you deploy your ingress pods by using a machine set on a smaller network, such as a `/27` or `/28`, you can simplify your load balancer targets.
```

Reviewer 2: Suggested editing for readability

+
[TIP]
====
You can list all IP addresses that exist in a network by checking the machine config pool's resources.
====

.Considerations

> * For a front-end IP address, you could use the same IP address for the front-end IP address, the Ingress Controller's load balancer and API load balancer. Check the vendor's documentation for this capability.

Reviewer 2: Per IBMSG, don't use "could":
Do not use in the present tense to mean might or can. Use only in the past conditional form, for example, “files that could not be identified”.

Replace with "can", or rewrite to something like:

"If your vendor supports X, you can..."

> * For a backend IP address, ensure that an IP address for an {product-title} control plane node does not change during the lifetime of the external load balancer. You can achieve by completing one of the following actions:

Reviewer 2: ```suggestion
* For a back-end IP address, ensure that an IP address for an {product-title} control plane node does not change during the lifetime of the external load balancer. You can achieve this by completing one of the following actions:
```

** Assign a static IP address to each control plane node.
** Configure each node to receive the same IP address from the DHCP every time the node requests a DHCP lease. Depending on the vendor, the DHCP lease might be in the form of an IP reservation or a static DHCP assignment.

> * For the Ingress Controller backend service, ensure you manually define each node that runs the Ingress Controller in the external load balancer. For example, if the Ingress Controller moves to an undefined node, an connection outage can occur.

Reviewer 2: ```suggestion
* Manually define each node that runs the Ingress Controller in the external load balancer for the Ingress Controller back-end service. For example, if the Ingress Controller moves to an undefined node, a connection outage can occur.
```

    server my-cluster-master-0 192.168.1.1020.2.3:22623 check inter 10s rise 2 fall 2
    server my-cluster-master-1 192.168.1.1030.2.1:22623 check inter 10s rise 2 fall 2

listen my-cluster-apps-443
        bind 192.168.1.100:443
        mode tcp
        balance roundrobin
    option httpchk
    http-check connect
    http-check send meth GET uri /healthz/ready
    http-check expect status 200
        server my-cluster-worker-0 192.168.1.111:443 check port 1936 inter 10s rise 2 fall 2
        server my-cluster-worker-1 192.168.1.112:443 check port 1936 inter 10s rise 2 fall 2
        server my-cluster-worker-2 192.168.1.113:443 check port 1936 inter 10s rise 2 fall 2

listen my-cluster-apps-80
        bind 192.168.1.100:80
        mode tcp
        balance roundrobin
    option httpchk
    http-check connect
    http-check send meth GET uri /healthz/ready
    http-check expect status 200
        server my-cluster-worker-0 192.168.1.111:80 check port 1936 inter 10s rise 2 fall 2
        server my-cluster-worker-1 192.168.1.112:80 check port 1936 inter 10s rise 2 fall 2
        server my-cluster-worker-2 192.168.1.113:80 check port 1936 inter 10s rise 2 fall 2

> #...

Reviewer 2: Not important, just think it looks tidier with the space, per the example shown here https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#signs-and-symbols-in-code-blocks

  "platform": "linux/amd64"
}
----
+
.. For the Machine config server resource, verify that cluster machine configuration API is accessible to:

> +

Reviewer 2: Similar comments to previous step, sentence doesn't feel finished ending on `to:` - for accessibility I'd tweak these a bit to make them feel more complete.

+
[source,terminal]
----
$ curl -I -L --insecure --resolve console-openshift-console.apps.<cluster_name>.<base_domain>:443:<Load Balancer Front End IP Address> https://console-openshift-console.apps.<cluster_name>.<base_domain>
----
+
If the configuration is correct, the output from the command shows the following response:
+
[source,json]
----
HTTP/1.1 200 OK
referrer-policy: strict-origin-when-cross-origin
set-cookie: csrf-token=UlYWOyQ62LWjw2h003xtYSKlh1a0Py2hhctw0WmV2YEdhJjFyQwWcGBsja261dGLgaYO0nxzVErhiXt6QepA7g==; Path=/; Secure; SameSite=Lax
x-content-type-options: nosniff
x-dns-prefetch-control: off
x-frame-options: DENY
x-xss-protection: 1; mode=block
date: Wed, 04 Oct 2023 16:29:38 GMT
content-type: text/html; charset=utf-8
set-cookie: 1e2670d92730b515ce3a1bb65da45062=1bf5e9573c9a2760c964ed1659cc1673; path=/; HttpOnly; Secure; SameSite=None
cache-control: private
----

. Configure the DNS records for your cluster to target the front-end IP addresses of the external load balancer. You must update records to your DNS server for the cluster API and applications over the load balancer.
+

> .Example of modified DNS records

Reviewer 2: ```suggestion
.Examples of modified DNS records
```

+
[source,dns]
----
<load_balancer_ip_address>  A  api.<cluster_name>.<base_domain>
A record pointing to Load Balancer Front End
----
+
[source,dns]
----
<load_balancer_ip_address>   A apps.<cluster_name>.<base_domain>
A record pointing to Load Balancer Front End
----
+
[IMPORTANT]
====
DNS propagation might take some time for each DNS record to become available. Ensure that each DNS record propagates before validating each record.
====

> . From the CLI, use `curl` to verify that the external load balancer and DNS record configuration are operational.

Reviewer 2: ```suggestion
. Use the `curl` CLI command to verify that the external load balancer and DNS record configuration are operational:
```

+
.. Verify that you can access the cluster API:
+
[source,terminal]
----
$ curl https://api.<cluster_name>.<base_domain>:6443/version --insecure
----
+
If the configuration is correct, you receive a JSON object in response:
+
[source,json]
----
{
  "major": "1",
  "minor": "11+",
  "gitVersion": "v1.11.0+ad103ed",
  "gitCommit": "ad103ed",
  "gitTreeState": "clean",
  "buildDate": "2019-01-09T06:44:10Z",
  "goVersion": "go1.10.3",
  "compiler": "gc",
  "platform": "linux/amd64"
  }
----
+

> .. Verify that you can access the cluster machine configuration API:

Reviewer 2: Same comment as before regarding suggested sentence structure for these type of steps

+
[source,terminal]
----
$ curl -v https://api.<cluster_name>.<base_domain>:22623/healthz --insecure
----
+
If the configuration is correct, the output from the command shows the following response:
+
[source,json]
----
HTTP/1.1 200 OK
Content-Length: 0
----
+

> .. Verify that you can access each cluster application on port 80:

Reviewer 2: Same comment as before regarding suggested sentence structure for these type of steps

set-cookie: 1e2670d92730b515ce3a1bb65da45062=9b714eb87e93cf34853e87a92d6894be; path=/; HttpOnly; Secure; SameSite=None
cache-control: private
----
+

> .. Verify that you can access each cluster application on port 443:

Reviewer 2: Same comment as before regarding suggested sentence structure for these type of steps
